### PR TITLE
Improved error proliferation through attributes

### PR DESCRIPTION
**Pull request by vortexofdoom**
First stab at addressing #31 

Fixed a couple typos/edited for readability/consistency in the `bitenum` module but didn't do anything with errors.

All errors I could figure out how to trigger now point at least in the general direction of the error in question.

Could have maybe pulled out the return into its own function but I didn't think it was worthwhile, it'd probably still be too long to comfortably sit on one line.

I did notice that you can have a field like this:
```
#[bit(1, rw)]
foo: [u3; 2]
```
and it won't throw an error at all. Unless I'm misunderstanding how the arrays work, that seems problematic, but beyond the scope of this issue maybe.

Could maybe be made cleaner by incorporating more `syn` types into the control flow, but I didn't want to mess with it too much.

### CODE REVIEW

**Comment by danlehmann** [comment]
Hey, thank you so much for this patch. Unfortunately, this is going to collide with https://github.com/danlehmann/bitfield/pull/32 , which I hope to land today. One of us will have to rebase :(

**Comment by vortexofdoom** [comment]
I should rebase it into one commit anyway, given what I changed in the others, so I can do that.

**Comment by danlehmann** [comment]
Thank you! I really hate to ask this of you, but I think today was just unfortunate timing.

**Comment by vortexofdoom** [comment]
I guess the wait would be a good excuse to expand into the main issue I saw with `bitenum`, namely, that the field is required to be an `Option` but the generated function creates a `Result`.

 This creates an odd error where it's not clear where the `Result` is coming from when debugging,  which is why I didn't mess with that much, probably requires a separate discussion, though I think it'd be in scope of this fix.

**Comment by danlehmann** [comment]
I saw you're making some formatting changes. I ran rustfmt over the whole project: https://github.com/danlehmann/bitfield/pull/36

Please let me know if you'd rather have that in before this patch or if you'd prefer to work from the current state

**Comment by vortexofdoom** [comment]
You're fine to merge it

**Comment by vortexofdoom** [comment]
All right @danlehmann, I made a bit of a mess of it, but I think I got everything hooked back up correctly.

Not much has changed overall, though I had to make `parse_field()` return a `Result` to pipe the errors through to the proc macro, and `Err(syn::Error::new_spanned(...).to_compile_error()` is starting to get ugly, but it works.

**Comment by danlehmann** [comment]
> `#[bit(1, rw)] foo: [u3; 2]`

I just tried (before your patch) and am getting this error:

message: bitfield!: Field foo has type u3, which doesn't match the number of bits (1) that are being used for it

File: bitbybit/src/bitfield.rs [hunk 1]
```diff
@@ -584,58 +600,140 @@ fn parse_field(base_data_size: usize, field: &&Field) -> FieldDefinition {
...
+                    let start = range_elements[0].parse::<usize>()
+                        .map_err(|_| if range_elements[0] == "" {
+                            syn::Error::new_spanned(
+                                &range_span,
+                                "bitfield!: Expected valid integer, but found an empty string"
+                            ).to_compile_error()
+                        } else {
+                            syn::Error::new_spanned(
+                                &range_span,
+                                format!("bitfield!: Expected valid range, e.g. 1..=8 but '{}' is not a number", range_elements[0])
```
Comment (danlehmann) [inline bitbybit/src/bitfield.rs:655]:
```suggestion
                                format!("bitfield!: Expected range (e.g. 1..=8). Saw '{}', which is not a valid", range_elements[0])
```

File: bitbybit/src/bitfield.rs [hunk 2]
```diff
@@ -584,58 +600,140 @@ fn parse_field(base_data_size: usize, field: &&Field) -> FieldDefinition {
...
                     if bit_index >= base_data_size {
-                        panic!("bitfield!: Bit index {} is out of range", bit_index);
+                        return Err(syn::Error::new_spanned(
+                            &range_span,
+                            format!("bitfield!: Bit index {} is out of range", bit_index),
+                        )
+                        .to_compile_error());
                     }
-                    range = Some(Range { start: bit_index, end: bit_index + 1 });
+                    range = Some(Range {
```
Comment (danlehmann) [inline bitbybit/src/bitfield.rs:733]: did you run this through some formatter other than rustfmt or through rustfmt with custom settings?

**Comment by danlehmann** [comment]
Thanks a lot for your contribution!

File: bitbybit/src/bitfield.rs (hunk 2 above)
Comment (danlehmann) [reply bitbybit/src/bitfield.rs:733]: Nevermind this!

**Review by danlehmann — APPROVED** [review]

**Comment by danlehmann** [comment]
Tested it and works great. This is a big quality of life improvement for users. Thanks again!

**Comment by danlehmann** [comment]
Converted a couple more here: https://github.com/danlehmann/bitfield/pull/37

**Pull request merged**